drivers/stepper/adi_tmc/adi_tmc5160_stepper_controller.c

```diff
@@ -0,0 +1,670 @@
+	return 0;
+}
+
+static int tmc5160_stepper_move(const struct device *dev, const int32_t steps)
+{
+	const struct tmc5160_stepper_config *config = dev->config;
+	struct tmc5160_stepper_data *data = dev->data;
+	int err;
+
+	if (config->is_sg_enabled) {
+		err = stallguard_enable(dev, false);
+		if (err != 0) {
+			return -EIO;
+		}
+	}
+
+	int32_t position;
+
+	err = stepper_get_actual_position(dev, &position);
+	if (err != 0) {
+		return -EIO;
+	}
+	int32_t target_position = position + steps;
+
+	err = tmc5160_write(config->controller, TMC5160_RAMPMODE,
+			    TMC5160_RAMPMODE_POSITIONING_MODE);
+	if (err != 0) {
+		return -EIO;
+	}
+	LOG_DBG("Stepper motor controller %s moved to %d by steps: %d", dev->name, target_position,
+		steps);
+	err = tmc5160_write(config->controller, TMC5160_XTARGET, target_position);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	if (config->is_sg_enabled) {
+		k_work_reschedule(&data->stallguard_dwork,
+				  K_MSEC(config->sg_velocity_check_interval_ms));
+	}
+#ifdef CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL
+	if (data->callback) {
+		k_work_reschedule(
+			&data->rampstat_callback_dwork,
+			K_MSEC(CONFIG_STEPPER_ADI_TMC5160_RAMPSTAT_POLL_INTERVAL_IN_MSEC));
+	}
+#endif
+	return 0;
+}
+
+static int tmc5160_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
+{
+	const struct tmc5160_stepper_config *config = dev->config;
+	uint32_t velocity_fclk;
+	int err;
+
+	calculate_velocity_from_hz_to_fclk(config->controller, velocity, &velocity_fclk);
+	err = tmc5160_write(config->controller, TMC5160_VMAX, velocity_fclk);
+	if (err != 0) {
+		LOG_ERR("%s: Failed to set max velocity", dev->name);
+		return -EIO;
+	}
+	return 0;
+}
+
+static int tmc5160_stepper_set_micro_step_res(const struct device *dev,
+					      enum stepper_micro_step_resolution res)
+{
+	const struct tmc5160_stepper_config *config = dev->config;
+	uint32_t reg_value;
+	int err;
+
+	err = tmc5160_read(config->controller, TMC5160_CHOPCONF, &reg_value);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	reg_value &= ~TMC5160_CHOPCONF_MRES_MASK;
+	reg_value |= ((MICRO_STEP_RES_INDEX(STEPPER_MICRO_STEP_256) - LOG2(res))
+		      << TMC5160_CHOPCONF_MRES_SHIFT);
+
+	err = tmc5160_write(config->controller, TMC5160_CHOPCONF, reg_value);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	LOG_DBG("Stepper motor controller %s set micro step resolution to 0x%x", dev->name,
+		reg_value);
+	return 0;
+}
+
+static int tmc5160_stepper_get_micro_step_res(const struct device *dev,
+					      enum stepper_micro_step_resolution *res)
+{
+	const struct tmc5160_stepper_config *config = dev->config;
+	uint32_t reg_value;
+	int err;
+
+	err = tmc5160_read(config->controller, TMC5160_CHOPCONF, &reg_value);
+	if (err != 0) {
+		return -EIO;
+	}
+	reg_value &= TMC5160_CHOPCONF_MRES_MASK;
+	reg_value >>= TMC5160_CHOPCONF_MRES_SHIFT;
+	*res = (1 << (MICRO_STEP_RES_INDEX(STEPPER_MICRO_STEP_256) - reg_value));
```

Comment:
can you do `*res = BIT(MICRO_STEP_RES_INDEX(STEPPER_MICRO_STEP_256) - reg_value)`?